File: leetCode/two pointers/SortColors.java
import java.util.*;

public class SortColors {
    public static void sortColors2(int[] colors, int k) {
        int min = 1, max = k;
        int i = 0;
        int pl = 0, pr = colors.length - 1;
        while (min < max) {
            while (i <= pr) {
                if (colors[i] == min) {
                    swap(colors, i, pl);
                    pl++;
                    i++;  
                    /* 这里i++, 是因为i从pl开始, 那么swap后, 
                    swap过来到i位置上的这个值, 在最开始的时候
                    已经验证过了, 所以无需再次验证. 如果i在swap
                    之后不变, 那么在有一种情况, 就是在最初始的
                    时候, i == pl == min, 此时swap, swap后, 
                    pl++, 但是i仍指向min, 这时候会再次swap, 
                    就打乱了顺序. 
                    */
                } else if (colors[i] == max) {
                    swap(colors, i, pr);
                    pr--;
                    /* 这里则不应该i++, 因为pr所指的值是未知的,
                    所以swap过来后需要验证一次.

                    */
                } else {
                    i++;
                }
            }
            i = pl;
            min++;
            max--;
        }
    }

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void main(String[] args) {
        int[] arr = { 3, 4, 1, 2, 7, 3, 5};
        sortColors2(arr, 7);
        System.out.println(Arrays.toString(arr));
    }
}

File: leetCode/High Frequncy/SortColors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    let colors = [0, 0, 0];
    let res = [];
    for (let num of nums) {
        if (num === 0) {
            colors[0]++;
        } else if (num === 1) {
            colors[1]++;
        } else {
            colors[2]++;
        }
    }
    let color = 0;
    for(let i = 0; i < nums.length; i++) {
        while (colors[color] === 0) color++;
        nums[i] = color;
        colors[color]--;
    }
};

// Solution II
const swap = (arr, i, j) => {
    let temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

const sortColorsII = (nums) => {
    let max = Math.max.apply(undefined, nums), min = Math.min.apply(undefined, nums);
    let left = 0, right = nums.length - 1;
    let i = left;
    while (min < max) {
        while (i <= right) {
            if (nums[i] === max) {
                swap(nums, i, right);
                right--;
            } else if (nums[i] === min) {
                swap(nums, i, left);
                left++;
                i++;
            } else i++;
        }
        i = left;
        min++;
        max--;
    }
}